##### Assignment 5 - Linear Regression

## Part 1 - Read your data frame.
The CSV file is "fifa_data.csv"

In [1]:
# Your code here
import pandas as pd
data = pd.read_csv("fifa_data.csv")

print(data.info())

data.drop(labels=["Loaned From", "Club", "Joined"], axis='columns', inplace=True)
data.dropna(axis='index', how='any', inplace=True)
print(data)

data.hist(bins=20)
print(data.head(n=10))
data.corr()

print(data.keys())
print(data.info())
print()
print(data["Nationality"].value_counts())

from pandas.plotting import scatter_matrix


def euro_converter(string):
    unit = string[0:1]
    value = string[1:len(string)-1]
    multiplier = string[-1]
    if multiplier =="0": return 0
    
    # print(unit, value, multiplier)
    value = float(value)
    if multiplier=="M": value *= 1000000
    if multiplier=="K": value *= 1000
    return value

data["Numeric Wage"] = data["Wage"].map(euro_converter)

compare = ["Overall", "Potential", "HeadingAccuracy", "Age", "Skill Moves", "Numeric Wage"]
scatter_matrix(data[compare])



<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18207 entries, 0 to 18206
Data columns (total 89 columns):
Unnamed: 0                  18207 non-null int64
ID                          18207 non-null int64
Name                        18207 non-null object
Age                         18207 non-null int64
Photo                       18207 non-null object
Nationality                 18207 non-null object
Flag                        18207 non-null object
Overall                     18207 non-null int64
Potential                   18207 non-null int64
Club                        17966 non-null object
Club Logo                   18207 non-null object
Value                       18207 non-null object
Wage                        18207 non-null object
Special                     18207 non-null int64
Preferred Foot              18159 non-null object
International Reputation    18159 non-null float64
Weak Foot                   18159 non-null float64
Skill Moves                 18159 non-null fl

    Unnamed: 0      ID               Name  Age  \
0            0  158023           L. Messi   31   
1            1   20801  Cristiano Ronaldo   33   
2            2  190871          Neymar Jr   26   
4            4  192985       K. De Bruyne   27   
5            5  183277          E. Hazard   27   
6            6  177003          L. Modrić   32   
7            7  176580          L. Suárez   31   
8            8  155862       Sergio Ramos   32   
10          10  188545     R. Lewandowski   29   
11          11  182521           T. Kroos   28   

                                             Photo Nationality  \
0   https://cdn.sofifa.org/players/4/19/158023.png   Argentina   
1    https://cdn.sofifa.org/players/4/19/20801.png    Portugal   
2   https://cdn.sofifa.org/players/4/19/190871.png      Brazil   
4   https://cdn.sofifa.org/players/4/19/192985.png     Belgium   
5   https://cdn.sofifa.org/players/4/19/183277.png     Belgium   
6   https://cdn.sofifa.org/players/4/19/177003.png   

array([[<matplotlib.axes._subplots.AxesSubplot object at 0x000001B0430BED30>,
      dtype=object)

## Part 2 - Display info
Take a look at the information.  Note any features that are missing values.  Note any features that are objects, but could be numerical

In [4]:
data.info()
data["LF"].value_counts()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 14743 entries, 0 to 18206
Data columns (total 87 columns):
Unnamed: 0                  14743 non-null int64
ID                          14743 non-null int64
Name                        14743 non-null object
Age                         14743 non-null int64
Photo                       14743 non-null object
Nationality                 14743 non-null object
Flag                        14743 non-null object
Overall                     14743 non-null int64
Potential                   14743 non-null int64
Club Logo                   14743 non-null object
Value                       14743 non-null object
Wage                        14743 non-null object
Special                     14743 non-null int64
Preferred Foot              14743 non-null object
International Reputation    14743 non-null float64
Weak Foot                   14743 non-null float64
Skill Moves                 14743 non-null float64
Work Rate                   14743 non-null o

61+2    653
62+2    617
64+2    614
63+2    612
59+2    605
60+2    605
58+2    591
65+2    576
57+2    544
66+2    540
56+2    504
55+2    487
68+2    471
54+2    467
67+2    460
53+2    417
52+2    386
51+2    380
69+2    380
50+2    326
70+2    317
49+2    297
71+2    289
72+2    258
48+2    247
47+2    242
46+2    235
45+2    220
73+2    196
44+2    184
       ... 
58+3      6
72+3      6
55+3      4
86+3      4
85+3      4
87+3      4
57+3      4
59+3      3
62+3      3
65+3      3
83+2      3
29+2      3
27+2      2
63+3      2
48+3      2
68+3      2
60+3      2
56+3      2
53+3      1
52+3      1
90+3      1
87+5      1
84+2      1
51+3      1
88+3      1
89+3      1
93+2      1
80+4      1
47+3      1
43+3      1
Name: LF, Length: 102, dtype: int64

### _Notes_
value and wage are objects that could be ints or floats
same with these
Height                      14743 non-null object
Weight                      14743 non-null object
LS                          14743 non-null object
ST                          14743 non-null object
RS                          14743 non-null object
LW                          14743 non-null object
LF                          14743 non-null object
CF                          14743 non-null object
RF                          14743 non-null object
RW                          14743 non-null object
LAM                         14743 non-null object
CAM                         14743 non-null object
RAM                         14743 non-null object
LM                          14743 non-null object
LCM                         14743 non-null object
CM                          14743 non-null object
RCM                         14743 non-null object
RM                          14743 non-null object
LWB                         14743 non-null object
LDM                         14743 non-null object
CDM                         14743 non-null object
RDM                         14743 non-null object
RWB                         14743 non-null object
LB                          14743 non-null object
LCB                         14743 non-null object
CB                          14743 non-null object
RCB                         14743 non-null object
RB

## Part 3a - Remove columns (features)
Refer to data_cleaning in the DataFrames directory

The "Loaned From" data feature is sparse and does not help us.

Remove that column from your data frame.

In [ ]:
# this part is already done on part 1. see above code for this part. 

#data.drop(label = ["Loaned From"]axis = "column")

## Part 3b - Remove rows (data instances)
Refer to data_cleaning in the DataFrames directory

For now, go ahead and remove any instance that is missing any data value.

In [ ]:
# Your code here

## Part 3c - We want to predict the wages for a player
But the wages in the data frame are not a numerical type.  I have created a handy function that will take the amount and return a numerical value.  Create a new column by mapping the euro_converter function on the "Wage" column.

In [ ]:
def euro_converter(string):
    unit = string[0:1]
    value = string[1:len(string)-1]
    multiplier = string[-1]
    if multiplier =="0": return 0
    
    # print(unit, value, multiplier)
    value = float(value)
    if multiplier=="M": value *= 1000000
    if multiplier=="K": value *= 1000
    return value

In [ ]:
# Your code here

## Part 4 - Split the data frame

Use the train_test_split() function to split the data set into training(75%) and test(25%) sets. 

In [ ]:
# Your code here

## Part 5 - Do some scatter plots
Take a look at "Potential", "Overall", and the new wages feature you created.

In [ ]:
# Your code here

## Part 6 - Linear Regression
1. Pull out "Overall" for the X and the wage feature you created for y. 
2. Fit the data.
3. Show R2 and Mean Square Error
4. Discuss the results

In [ ]:
# Your code here

Discussion: 

## Part 7 - Add features to the Model
1. Pick up to 3 other features that you think might improve the model and use them for X.
2. Fit the data
3. Show the scores
4. Discuss the results

In [ ]:
# Your code here

Discussion: 

## Part 8 - Test Set time
Evaluate the model on the test set and discuss your results.

In [ ]:
# Your code here

Discussion:

# Bonus options
For each of the options, redo your regression using the new features, report the results and discuss.


1. Do a better job of filtering the data.  It looks like there are 18159 instances that have numerical performance values.  Filter unneeded columns so that those 18159 instances remain. 
2. Look at the other columns that contain objects and see if any could be useful in your regression.  Write a map function and apply.  
2. Add in polynomial features.